doc/C/timidity.1
.TH timidity 1 "April 25 2004" "2.13.0"
.SH NAME
TiMidity++ \- MIDI\-to\-WAVE converter and player
.SH SYNOPSIS
.B timidity
[options] filename [...]
.SH DESCRIPTION
\fBTiMidity++\fP is a converter that converts some of MIDI files
(supported formats: Standard MIDI files (*.mid), Recomposer files
(*.rcp, *.r36, *.g18, *.g36) and Module files (*.mod)) into formatted
audio files (e.g. RIFF WAVE).  \fBTiMidity++\fP uses Gravis
Ultrasound\-compatible patch files or Soundfonts (*.sfx, *.sf2) to
generate digital audio data from MIDI files.  The digital audio data
generated by \fBTiMidity++\fP can be stored in a file for processing,
or played in real time through an audio device.
.br
In real time playing, \fBTiMidity++\fP can show the lyrics contained
in KAR or WRD files.
.SH FILENAME
You can use the following expressions as the \fIfilename\fP argument:
.TP
.B \-
Read a MIDI file from standard input.
.TP
.B path/filename
Read a MIDI file from the specified path on a filesystem.
.TP
.B dir:directory
.br
.ns
.TP
.B directory/
Read and play all MIDI files in the specified \fIdirectory\fP.  For
example,
.sp
% timidity some/where/
.sp
plays all files in the directory \fBsome/where/\fP.
.TP
.B Archive File
Extract and play the file(s) in the archive.  If you want to specify a
certain MIDI file in the archive, append \fB#<MIDI\-filename>\fP to
the archive name.  The path after `#' allows the use of the wildcard
expressions (case insensitive).
.br
You can use escape sequence \fB\\xHH\fP, where `HH' is a ASCII number
in hexadecimal integer.
.sp
For example:
.RS
.TP
% timidity file.zip#file.mid
Plays \fBfile.mid\fP in \fBfile.zip\fP
.TP
% timidity file.lzh#*.mid
Plays any files that match the wildcard expression \fB*.mid\fP in
\fBfile.lzh\fP
.TP
% timidity file.tgz#*
This expression is the same as \fBfile.tgz\fP
.LP
Since these mechanism are contained in \fBTiMidity++\fP itself, you
can use this syntax even in the MS Windows environment.
.sp
\fBTiMidity++\fP can handle the following archive formats:
.TP
.B tar (*.tar)
.TP
.B tar+gzip (*.tar.gz, *.tgz)
.TP
.B zip (*.zip)
.TP
.B lzh (*.lzh, *.lha)
(lh0, lh1, lh2, lh3, lh4, lh5, lh6, lz4, lzs and lz5 are available)
.LP
Other archives can be expanded if expander command is specified on the
compile phase.  \fBTiMidity++\fP will pipe that command.
.RE
.TP
.B news://news\-server[:port]/Message\-ID
.br
.ns
.TP
.B news://news\-server[:port]/newsgroup[/first\-last]
Play the MIDI file in the specified article on the news server.  If a
newsgroup is specified \fBTiMidity++\fP plays all MIDI files found in
any article posted to that newsgroup.
.br
\fBTiMidity++\fP parses MIME Multi\-part messages in case of
\fBnews://*\fP scheme, extracts MIDI file that was post to that group,
or in case of normal file, you can explicitly handle MIME documents by
naming that file with *.mime extension, or prefix that file with
\fBmime:\fP
.sp
The following MIME\-types are allowed:
.RS
.TP
.B uu\-encoded file
\fBbegin\fP
.br
is required
.TP
.B base64 encoded
\fBContent\-Transfer\-Encoding: base64\fP
.br
is required
.TP
.B quoted\-string
\fBContent\-Transfer\-Encoding: quoted\-string\fP
.br
is required
.TP
.B Mac BinHex format
only HQX format is available
.RE
.TP
.B http://address
.br
.ns
.TP
.B ftp://address
Play the file specified in the URL.
.sp
For example:
.RS
.TP
% timidity http://www.goice.co.jp/member/mo/dist/midi/impromptu.mid
plays the specified MIDI\-file directly from the network.
.LP
If these expression are used in the *.cfg files, you can even use
patch\-files (and others) from remote machines.
.RE
.SH INPUT FILE
\fBTiMidity++\fP can handle the following file formats:
.TP
.B .mid, .rmi (Format 0, 1, 2)
Standard MIDI File
.TP
.B .rcp, .r36, .g18, .g36 (Recomposer formats)
Recomposer format which is product for COME ON MUSIC co.
.TP
.B .mfi (MFi Version 3 \- Melody Format for i\-Mode)
i\-Mode is Japanese local mobile phone
.TP
.B .kar (Karaoke format)
Displays the lyrics as a Lyric Meta Event message.
.TP
.B .mod, mod.* (Module file)
.TP
.B .wrd (WRD format)
.SH OPTIONS
The following command line options are accepted by \fBTiMidity++\fP:
.TP
.BI "\-A " [n][,m](a)
.br
.ns
.TP
.BI \-\-amplification= n
.br
.ns
.TP
.BI \-\-drumpower= m
.br
.ns
.TP
.B \-\-[no\-]volume\-compensation
Multiplies the master volume by \fIn\fP%.  Default value is 70%.
Higher amplification makes louder sounds.  You can specify the drum
power, ratio of drum volume from the other channels.  The allowed
values of amplification range from \fB0\fP (no sound) to \fB800\fP.
.br
Optionally to put `a' character along with \fB\-a\fP option, or to use
\fB\-\-volume\-compensation\fP, instructs \fBTiMidity++\fP to
regularize the volume.  You can easily gain dynamic range.
.sp
For example:
.RS
.TP
.B \-A90
volume 90%, drum power 100%, compensation is off
.TP
.B \-A,120
volume 70%, drum power 120%, compensation is off
.TP
.B \-A90,120
volume 90%, drum power 120%, compensation is off
.TP
.B \-Aa
volume 70%, drum power 100%, compensation is on
.TP
.B \-A90a
volume 90%, drum power 100%, compensation is on
.TP
.B \-A,120a
volume 70%, drum power 120%, compensation is on
.TP
.B \-A90,120a
volume 90%, drum power 120%, compensation is on
.RE
.TP
.B \-a, \-\-[no\-]anti\-alias
Turns on antialiasing.  Samples are run through a lowpass filter
before playing, which reduces aliasing noise at low resampling
frequencies.
.TP
.BI "\-B " n,m ", \-\-buffer\-fragments=" n,m
For the Linux/FreeBSD/OSS/ALSA/Windows sound driver, selects the
number of buffer fragments in interactive mode.  Increasing the number
of fragments may reduce choppiness when many processes are running.
It will make \fBTiMidity++\fP seem to respond sluggishly to fast
forward, rewind, and volume controls, and it will throw the status
display off sync.  Specify a fragments number of 0 to use the maximum
number of fragments available.
.TP
.BI "\-C " n ", \-\-control\-ratio=" n
Sets the ratio of sampling and control frequencies.  This determines
how often envelopes are recalculated \-\- small ratios yield better
quality but use more CPU time.
.TP
.BI "\-c " file ", \-\-config\-file=" file
Reads an extra configuration file.
.TP
.BI "\-D " n ", \-\-drum\-channel=" n
Marks channel as a drum channel.  If channel is negative, channel
\-\fIn\fP is marked as an instrumental channel.  If \fIn\fP is
\fB0\fP, all channels are marked as instrumental.
.TP
.BI "\-d " dir ", \-\-interface\-path=" dir
Specifies the directory containing installed dynamic\-link interface
modules.
.TP
.BI "\-E " mode ", \-\-ext=" mode
Set \fBTiMidity++\fP extend modes.  The following modes are available
(capitalized switch means disable this feature):
.RS
.TP
.B w/W, \-\-[no\-]mod\-wheel
Enable/disable modulation controlling.
.TP
.B p/P, \-\-[no\-]portamento
Enable/disable portamento controlling.
.TP
.B v/V, \-\-[no\-]vibrato
Enable/disable NRPM vibration.
.TP
.B s/S, \-\-[no\-]ch\-pressure
Enable/disable channel pressure controlling.
.TP
.B e/E, \-\-[no\-]mod\-envelope
Enable/disable modulation envelope controlling.
.TP
.B t/T, \-\-[no\-]trace\-text\-meta
Enable/disable tracing all Text Meta Events.
.TP
.B o/O, \-\-[no\-]overlap\-voice
Accept/reject pronouncing multiple same notes.
.TP
.B z/Z, \-\-[no\-]temper\-control
Enable/disable MIDI Tuning Standard temperament controlling.
.TP
.BI m HH ", \-\-default\-mid=" HH
Sets the manufacturer ID to \fIHH\fP (where \fIHH\fP are two
hex\-digits).
.br
\fIHH\fP values of \fBGS/gs\fP, \fBXG/xg\fP or \fBGM/gm\fP are
understood as \fB41\fP, \fB43\fP and \fB7e\fP respectively.
.TP
.BI M HH ", \-\-system\-mid=" HH
Sets the system manufacturer ID to \fIHH\fP (where \fIHH\fP are two
hex\-digits).
.br
In this option, the manufacuture ID is set unchangable. Manufacuture
ID from the input file would be ignored.
.TP
.BI b n ", \-\-default\-bank=" n
Use tone bank \fIn\fP as the default.
.TP
.BI B n ", \-\-force\-bank=" n
Sets the bank number of all channels to \fIn\fP.
.TP
.BI i n[/m] ", \-\-default\-program=" n[/m]
Use the program number as the default instrument.  Any Program Change
events in MIDI files will override this option.
.br
If \fIn\fP is followed by \fI/m\fP the default program number of the
channel \fIm\fP is specified by \fIn\fP.
.TP
.BI I n[/m] ", \-\-force\-program=" n[/m]
Similar to \fB\-Ei\fP but this ignores all program changes.
.TP
.BI "F " args
For effects.  See below.  In \fIargs\fP option, you can specify
following effect options:
.RS
.TP
.BI delay= (d|l|r|b)[,msec] ", \-\-delay=" (d|l|r|b)[,msec]
Sets delay type.
.RS
.TP
.B d, 0
Disabled delay effect.
.TP
.B l, 1
Left delay.
.TP
.B r, 2
Right delay.
.TP
.B b, 3
Swap left & right.
.LP
Optional \fImsec\fP is the delay time.
.RE
.TP
.BI chorus= (d|n|s)[,level] ", \-\-chorus=" (d|n|s)[,level]
.RS
.TP
.B d, 0
Disable this effect.
.TP
.B n, 1
Enable MIDI chorus effect control.
.TP
.B s, 2
Surround sound, chorus detuned to a lesser degree (default).
.LP
The optional parameter \fIlevel\fP specifies the chorus level \fB0\fP
to \fB127\fP.
.RE
.TP
.BI reverb= (d|n|g)[,level] ", \-\-reverb=" (d|n|g)[,level]
.TP
.BI reverb= (f|G)[,level,scaleroom,offsetroom,predelay]
.TP
.BI "\-\-reverb=" (f|G)[,level,scaleroom,offsetroom,predelay]
.RS
.TP
.B d, 0
Disable MIDI reverb effect control.
.TP
.B n, 1
Enable MIDI reverb effect control.  This effect is only available in
stereo.
.TP
.B g, 2
Global reverb effect.
.TP
.B f, 3
Freeverb MIDI reverb effect control (default).
.TP
.B G, 4
Global Freeverb effect.
.LP
The optional parameter \fIlevel\fP specifies the reverb level \fB0\fP
to \fB127\fP.
.br
For Freeverb effect, additional parameters \fIscaleroom\fP,
\fIoffsetroom\fP and \fIpredelay\fP can be specified.
\fIscaleroom\fP and \fIoffsetroom\fP specify variables for the room
size calculated as
.br
roomsize = C * \fIscaleroom\fP + \fIoffsetroom\fP
.br
where C is the reverberation character.
\fIpredelay\fP specifies the factor for pre-delay time of
reverberation in percent.
Any of the optional parameters can be omitted.
Default values are \fB0.28\fP, \fB0.7\fP and \fB100\fP respectively.
.RE
.TP
.BI vlpf= (d|c|m) ", \-\-voice\-lpf=" (d|c|m)
.RS
.TP
.B d, 0
Disable LPF effect.
.TP
.B c, 1
Chamberlin resonant LPF (12dB/oct) (default).
.TP
.B m, 2
Moog resonant low\-pass VCF (24dB/oct)
.RE
.TP
.BI ns= n ", \-\-noise\-shaping=" n
Enable the \fIn\fP th degree noiseshaping filter.  The distortion at
decay stage is improved, but the noise on human auditory feeling
increases because it shifts to a high frequency.  In case of 8\-bit
linear encoding, valid values of \fIn\fP are in the interval from
\fB0\fP (min) to \fB4\fP (max).  Default value is \fB4\fP.  In case of
16\-bit linear encoding, valid values of n are in the interval from
\fB0\fP to \fB4\fP.  According to the value, it works as following.
Default value is \fB4\fP.
.RS
.TP
.B 0
No noise shaping.
.TP
.B 1
Traditional noise shaping.
.TP
.B 2
Overdrive-like soft-clipping + new noise shaping.
.TP
.B 3
Tube-amplifier-like soft-clipping + new noise shaping.
.TP
.B 4
New noise shaping.
.RE
.TP
.BI resamp= (d|l|c|L|n|g) ", \-\-resample=" (d|l|c|L|n|g)
.RS
.TP
.B d, 0
No interpolation.
.TP
.B l, 1
Linear interpolation.
.TP
.B c, 2
Cubic spline interpolation.
.TP
.B L, 3
Lagrange method.
.TP
.B n, 4
Newton polynomial interpolation.
.TP
.B g, 5
Modified Gauss effect (default).
.LP
This option affects the behavior of \fB\-N\fP option.
.RE
.RE
.RE
.TP
.B \-e, \-\-evil
Make \fBTiMidity++\fP evil.  For the Win32 version, this increases the
task priority by one.  It can give better playback when you switch
tasks at the expense of slowing all other tasks down.
.TP
.B \-F, \-\-[no\-]fast\-panning
Turns on fast panning to accommodate MIDI pieces that expect panning
adjustments to affect notes that are already playing.  Some files that
don't expect this have a habit of flipping balance rapidly between
left and right, which can cause severe popping when the \fB\-F\fP flag
is used.
.br
In the current version of \fBTiMidity++\fP this option is a toggle.
.TP
.B \-f, \-\-[no\-]fast\-decay
Toggles fast envelopes.  This option makes \fBTiMidity++\fP faster but

interface/timidity.el
;; TiMidity interface for Emacs
;;
;;   TiMidity++ -- MIDI to WAVE converter and player
;;   Copyright (C) 1999-2002 Masanao Izumo <mo@goice.co.jp>
;;   Copyright (C) 1995 Tuukka Toivonen <tt@cgs.fi>
;;
;;   This program is free software; you can redistribute it and/or modify
;;   it under the terms of the GNU General Public License as published by
;;   the Free Software Foundation; either version 2 of the License, or
;;   (at your option) any later version.
;;
;;   This program is distributed in the hope that it will be useful,
;;   but WITHOUT ANY WARRANTY; without even the implied warranty of
;;   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;;   GNU General Public License for more details.
;;
;;   You should have received a copy of the GNU General Public License
;;   along with this program; if not, write to the Free Software
;;   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
;;
;;
;; Author: Masanao Izumo <mo@goice.co.jp>
;; Create: Tue Oct 7 1997
;; LastModified: Tue Dec 9 1997
;; Keywords: timidity
;;
;;; Commentary:
;;
;; You write follows in ~/.emacs
;; (autoload 'timidity "timidity" "TiMidity Interface" t)
;; (setq timidity-prog-path "/usr/local/bin/timidity")
;;
;; Then you type:
;; M-x timidity

;; Configuration parameters.
; Absolute path of timidity.
(defvar timidity-prog-path "/usr/bin/timidity")

; String list for timidity program options.
(defvar timidity-default-options nil)

;; Follows are shouldn't need to be changed.
(defvar timidity-drumchannels nil)
(defvar timidity-prev-files nil)
(defvar timidity-next-files nil)
(defvar timidity-cmd-buffer-name "*TiMidityCmd*")
(defvar timidity-cmd-buffer nil)
(defvar timidity-cmd-process nil)
(defvar timidity-main-buffer-name "*TiMidity*")
(defvar timidity-main-buffer nil)
(defvar timidity-msg-buffer-name "*TiMidityMsg*")
(defvar timidity-msg-buffer nil)
(defvar timidity-orig-window-configuration nil)
(defvar timidity-note-table-ofs nil)
(defvar timidity-error-flag nil)
(defvar timidity-total-time 0)
(defvar timidity-current-time-bar 0)
(defvar timidity-current-filename "")
(defvar timidity-main-volume 0)
(defvar timidity-reset-bufstr nil)
(defvar timidity-cmd-version nil)
(defvar timidity-playing-flag nil)
(defvar timidity-pre-byte-compiles
  '(timidity-cmd-filter
    timidity-CMSG timidity-CURT timidity-NOTE timidity-PROG timidity-VOL
    timidity-EXP timidity-PAN timidity-SUS timidity-PIT))
(defvar timidity-jump-event-flag nil)

(defvar timidity-mode-map nil)
(if timidity-mode-map
    nil
  (setq timidity-mode-map (make-sparse-keymap))
  (define-key timidity-mode-map "l" 'timidity-load-file)
  (define-key timidity-mode-map "L" 'timidity-load-file-list)
  (define-key timidity-mode-map "n" 'timidity-next-file)
  (define-key timidity-mode-map "p" 'timidity-prev-file)
  (define-key timidity-mode-map "q" 'timidity-quit)
  (define-key timidity-mode-map "d" 'timidity-select-drumchannel)
  (define-key timidity-mode-map "r" 'timidity-replay-cmd)
  (define-key timidity-mode-map "V" 'timidity-simple-cmd)
  (define-key timidity-mode-map "v" 'timidity-simple-cmd)
  (define-key timidity-mode-map "1" 'timidity-simple-cmd)
  (define-key timidity-mode-map "2" 'timidity-simple-cmd)
  (define-key timidity-mode-map "3" 'timidity-simple-cmd)
  (define-key timidity-mode-map "4" 'timidity-simple-cmd)
  (define-key timidity-mode-map "5" 'timidity-simple-cmd)
  (define-key timidity-mode-map "6" 'timidity-simple-cmd)
  (define-key timidity-mode-map "f" 'timidity-simple-cmd)
  (define-key timidity-mode-map "b" 'timidity-simple-cmd)
  (define-key timidity-mode-map " " 'timidity-simple-cmd)
  (define-key timidity-mode-map "+" 'timidity-simple-cmd)
  (define-key timidity-mode-map "-" 'timidity-simple-cmd)
  (define-key timidity-mode-map ">" 'timidity-simple-cmd)
  (define-key timidity-mode-map "<" 'timidity-simple-cmd)
  (define-key timidity-mode-map "O" 'timidity-simple-cmd)
  (define-key timidity-mode-map "o" 'timidity-simple-cmd)
  (define-key timidity-mode-map "g" 'timidity-simple-cmd)
  )

(defun timidity ()
  (interactive)
  (let (width height check)
    (if (timidity-check-run)
	(progn
	  (switch-to-buffer timidity-main-buffer-name)
	  (error "Already timidity is running.")))
    (setq timidity-orig-window-configuration (current-window-configuration))
    (setq timidity-main-buffer (switch-to-buffer timidity-main-buffer-name))
    (delete-other-windows)
    (setq width (window-width))
    (setq height (1- (window-height)))
    (if (or (< width 80) (< height 23))
	(error "Window is too small."))
    (kill-all-local-variables)
    (use-local-map timidity-mode-map)
    (setq mode-name "TiMidity")
    (setq major-mode 'timidity-mode)
    (run-hooks 'timidity-mode-hook)
    (timidity-make-windows)
    (timidity-demo)
    (timidity-run)))

(defun timidity-simple-modeline ()
  (setq mode-line-modified nil)
  (setq mc-verbose-code nil)
  (setq mc-flag nil)
  (make-local-variable 'line-number-mode)
  (setq line-number-mode nil)
  (setq mode-line-mc-status nil)
  (setq mode-line-buffer-identification '("  ---: %15b")))

(defun timidity-make-windows ()
  (split-window nil 20)
  (other-window 1)
  (setq timidity-cmd-buffer (switch-to-buffer timidity-cmd-buffer-name))
  (lisp-interaction-mode)
  (timidity-simple-modeline)
  (erase-buffer)
  (setq timidity-msg-buffer (switch-to-buffer timidity-msg-buffer-name))
  (other-window 1)
  (setq truncate-lines t)
  (setq buffer-read-only t)
  (timidity-simple-modeline))

(defun timidity-check-run ()
  (if (and timidity-cmd-process
	   (or (null (buffer-name (process-buffer timidity-cmd-process)))
	       (eq 'exit (process-status timidity-cmd-process))))
      (setq timidity-cmd-process nil))
  timidity-cmd-process)

(defun timidity-run ()
  (setq timidity-error-flag nil)
  (if (timidity-check-run)
      timidity-cmd-process
    (let ((type (if (boundp 'MULE) "mule" "emacs")))
      (setq timidity-cmd-process
	    (apply 'start-process
		   "TiMidity"
		   timidity-cmd-buffer
		   timidity-prog-path
		   (append timidity-default-options (list "-ietv" type))))
      (set-process-filter timidity-cmd-process
			  (symbol-function 'timidity-cmd-filter))
      (set-process-sentinel timidity-cmd-process 'timidity-cmd-sentinel)
      (process-kill-without-query timidity-cmd-process))))

(defun timidity-cmd-filter (process string)
  (or timidity-error-flag
      (save-excursion
	(set-buffer timidity-cmd-buffer)
	(goto-char (point-max))
	(insert string)
	(skip-chars-backward "^\n")
	(condition-case err
	    (progn
	      (eval-region (point-min) (point))
	      (delete-region (point-min) (point)))
	  (error
	   (setq timidity-error-flag t)
	   (timidity-quit)
	   (switch-to-buffer timidity-cmd-buffer-name)
	   (error "%s" err))))))

(defun timidity-cmd-sentinel (process event)
  (if timidity-error-flag
      (message "TiMidity Error")
    (message "TiMidity ... done"))
  (setq timidity-playing-flag nil)
  (setq timidity-cmd-process nil))

(defun timidity-VERSION (str)
  (setq timidity-cmd-version str))

(defun timidity-NEXT ()
  (if timidity-next-files
      (timidity-next-file)
    (setq timidity-playing-flag nil)
    (sit-for 1)
    (timidity-demo)))

(defun timidity-CMSG (type msg)
  (save-excursion
    (let (origw msgw)
      (set-buffer timidity-msg-buffer)
      (goto-char (point-max))
      (insert msg ?\n)
      (setq msgw (get-buffer-window timidity-msg-buffer))
      (if msgw
	  (progn
	    (setq origw (selected-window))
	    (if (eq msgw origw)
		nil
	      (select-window msgw)
	      (goto-char (point-max))
	      (select-window origw)))))))

(defun timidity-TIME (time)
  (setq timidity-total-time time))

(defun timidity-DRUMS (drums)
  (setq timidity-drumchannels drums))

(defun timidity-MVOL (mv)
  (setq timidity-main-volume mv))

(defun timidity-FILE (name)
  (setq timidity-current-filename name)
  (let ((f (timidity-expand-file-name name)))
    (or (member f timidity-next-files)
	(member f timidity-prev-files)
	(setq timidity-prev-files (cons name timidity-prev-files)))))

(defun timidity-CURT (secs v)
  (set-buffer timidity-main-buffer)
  (let ((bar (/ (* 48 secs) timidity-total-time))
	(buffer-read-only nil)
	p)
    (if (> bar 48) (setq bar 48))
    (if (= bar timidity-current-time-bar)
	nil
      (timidity-update-time-bar bar)
      (setq timidity-current-time-bar bar))
    (goto-char (point-min))
    (forward-line 2)
    (skip-chars-backward "^|")
    (delete-char 12)
    (insert (format "%2d:%02d /%2d:%02d"
		    (/ secs 60)
		    (% secs 60)
		    (/ timidity-total-time 60)
		    (% timidity-total-time 60)))))

(defun timidity-NOTE (ch note stat)
  (set-buffer timidity-main-buffer)
  (let ((buffer-read-only nil))
    (goto-char (+ timidity-note-table-ofs
		  (* ch 80)
		  3
		  (% note 48)))
    (delete-char 1)
    (insert (if (eq stat 1)
		(aref "cCdDefFgGaAb" (% note 12))
	      (aref "..+_," stat)))))

(defun timidity-PROG (ch val)
  (set-buffer timidity-main-buffer)
  (let ((buffer-read-only nil))
    (goto-char (+ timidity-note-table-ofs
		  (* ch 80)
		  53))
    (delete-char 3)
    (insert (format "%03d" val))))

(defun timidity-VOL (ch val)
  (set-buffer timidity-main-buffer)
  (let ((buffer-read-only nil))
    (goto-char (+ timidity-note-table-ofs
		  (* ch 80)
		  57))
    (delete-char 3)
    (insert (format "%3d" val))))

(defun timidity-EXP (ch val)
  (set-buffer timidity-main-buffer)
  (let ((buffer-read-only nil))
    (goto-char (+ timidity-note-table-ofs
		  (* ch 80)
		  61))
    (delete-char 3)
    (insert (format "%3d" val))))

(defun timidity-PAN (ch val)
  (set-buffer timidity-main-buffer)
  (let ((buffer-read-only nil))
    (goto-char (+ timidity-note-table-ofs
		  (* ch 80)
		  65))
    (delete-char 3)
    (cond ((= val -1)		(insert "   "))
	  ((< val 5)		(insert " L "))
	  ((> val 123)		(insert " R "))
	  ((and (> val 60) (< val 68)) (insert " C "))
	  (t
	   (setq val (/ (* 100 (- val 64)) 64))
	   (if (>= val 0)
	       (insert ?+)
	     (insert ?-)
	     (setq val (- val)))
	   (insert (format "%02d" val))))))

(defun timidity-SUS (ch val)
  (set-buffer timidity-main-buffer)
  (let ((buffer-read-only nil))
    (goto-char (+ timidity-note-table-ofs
		  (* ch 80)
		  69))
    (delete-char 1)
    (insert (if (zerop val) ?\  ?S))))

(defun timidity-PIT (ch val)
  (set-buffer timidity-main-buffer)
  (let ((buffer-read-only nil))
    (goto-char (+ timidity-note-table-ofs
		  (* ch 80)
		  71))
    (delete-char 1)
    (insert (cond ((= val -1) ?=)
		  ((> val 8192) ?>)
		  ((< val 8192) ?<)
		  (t ?\  )))))

(defun timidity-RESET ()
  (set-buffer timidity-main-buffer)
  (let ((buffer-read-only nil)
	i s s1)
    (sit-for 0) ; flush
    (erase-buffer)
    (if timidity-reset-bufstr
	(insert timidity-reset-bufstr)
      (insert ?\n)
      (insert "  |>                                               |            \n")
      (insert "Ch c d ef g a bc d ef g a bc d ef g a bc d ef g a b  Prg Vol Exp Pan S B\n")
      (setq s (make-string 48 ?.))
      (setq s1 (make-string 28 ?\ ))
      (setq i 0)
      (while (< i 16)
	(insert (format "%02d " (1+ i)) s s1 ?\n)
	(setq i (1+ i)))
      (setq timidity-reset-bufstr (buffer-string)))
    (setq timidity-current-time-bar 0)
    (goto-char (point-min))
    (insert "File: " timidity-current-filename)
    (forward-line 3)
    (setq timidity-note-table-ofs (point))
    (setq i 0)
    (while (< i 16)
      (if (timidity-is-drumchannel i)
	  (progn
	    (forward-char 2)
	    (delete-char 1)
	    (insert ?*)))
      (setq i (1+ i))
      (forward-line 1))))

(defun timidity-load-file (fname)
  (interactive "FMIDI file: ")
  (if fname
      (progn
	(setq timidity-playing-flag t)
	(timidity-run)
	(setq fname (timidity-expand-file-name fname))
	(garbage-collect)
	(process-send-string timidity-cmd-process
			     (format "L\nPLAY %s\n" fname))
	(message "MIDI File: %s" fname))
    (setq timidity-playing-flag nil)))

(defun timidity-load-file-list ()
  (interactive)
  (let ((f (concat (expand-file-name "") "/"))
	files)
    (while (not (string= "" (setq f (read-file-name
				     "MIDI File: "
				     (file-name-directory
				      (timidity-expand-file-name f))
				     ""))))
      (setq files (cons (timidity-expand-file-name f) files)))
    (setq timidity-next-files (append (reverse files) timidity-next-files))
    (timidity-next-file)))

(defun timidity-next-file ()
  (interactive)
  (if timidity-next-files
      (let ((f (car timidity-next-files)))
	(setq timidity-next-files (cdr timidity-next-files))
	(setq timidity-prev-files (cons f timidity-prev-files))
	(timidity-load-file f))))

(defun timidity-prev-file ()
  (interactive)
  (if timidity-playing-flag
      (if (cdr timidity-prev-files)
	  (let ((f (car timidity-prev-files)))
	    (setq timidity-prev-files (cdr timidity-prev-files))
	    (setq timidity-next-files (cons f timidity-next-files))
	    (timidity-load-file (car timidity-prev-files)))
	(timidity-load-file (car-safe timidity-prev-files)))
    (timidity-load-file (car-safe timidity-prev-files))))

(defun timidity-replay-cmd ()
  (interactive)
  (if timidity-playing-flag
      (timidity-simple-cmd 1 ?r)
    (timidity-load-file (car-safe timidity-prev-files))))

(defun timidity-quit ()
  (interactive)
  (if (and timidity-cmd-process
	   (or (null (buffer-name (process-buffer timidity-cmd-process)))
	       (eq 'exit (process-status timidity-cmd-process))))
      (setq timidity-cmd-process nil))
  (if timidity-cmd-process
      (progn
	(process-send-string timidity-cmd-process (format "QUIT\n"))
	(setq timidity-cmd-process nil)))
  (if timidity-orig-window-configuration
      (progn
	(set-window-configuration timidity-orig-window-configuration)
	(setq timidity-orig-window-configuration nil))))

(defun timidity-simple-cmd (&optional cnt cmd)
  (interactive "p")
  (process-send-string timidity-cmd-process
		       (format "%c%d\n"
			       (if cmd cmd last-command-char)
			       (if cnt cnt 1))))

(defun timidity-select-drumchannel (ch)
  (interactive "nChannel: ")
  (if (or (not (integerp ch))
	  (< ch 1)
	  (> ch 16))
      (error "Invalid channel: %s" ch))
  (process-send-string timidity-cmd-process (format "d%d\n" (1- ch))))

(defun timidity-is-drumchannel (ch)
  (not (zerop (logand timidity-drumchannels (lsh 1 ch)))))

(defun timidity-update-time-bar (bar)
  (let (diff)
    (setq diff (- bar timidity-current-time-bar))
    (goto-char (point-min))
    (forward-line 1)
    (forward-char 3)
    (if (< 0 diff)
	(progn
	  (insert-char ?= diff)
	  (skip-chars-forward "=")
	  (if (= 48 bar)
	      (delete-char diff)
	    (forward-char 1)
	    (delete-char diff)))
      (setq diff (- diff))
      (delete-char diff)
      (skip-chars-forward "=")
      (if (eq (following-char) ?|)
	  (progn
	    (insert-char ?> 1)
	    (setq diff (1- diff))))
      (forward-char 1)
      (insert-char ?\  diff))))

(defun timidity-expand-file-name (fname)
  (if (or (string-match "^http://" fname)
	  (string-match "^ftp://" fname)
	  (string-match "^file://" fname)
	  (string-match "^news://" fname)
	  (string-match "^mime:" fname))
      fname
    (expand-file-name fname)))

(or (fboundp 'member)
    (defun member (elt lis)
      "A member function for emacs 18."
      (let ((loop t))
	(while (and loop lis)
	  (if (equal elt (car lis))
	      (setq loop nil)
	    (setq lis (cdr lis))))
	lis)))

(defvar timidity-demo-string
'("#######   ###   #     #   ###   ######    ###   ####### #     #"
  "   #       #    ##   ##    #    #     #    #       #     #   # "
  "   #       #    # # # #    #    #     #    #       #      # #  "
  "   #       #    #  #  #    #    #     #    #       #       #   "
  "   #       #    #     #    #    #     #    #       #       #   "
  "   #       #    #     #    #    #     #    #       #       #   "
  "   #      ###   #     #   ###   ######    ###      #       #   "
  "              TiMidity emacs interface written by Masanao Izumo"
  "                                                               "
  "   l/L: Load & Play a MIDI file                                "
  "   n/p: Next/Prev file                                         "
  "   r  : Replay a MIDI file                                     "
  "   q: Quit                                                     "
))

(defun timidity-demo ()
  (interactive)
  (set-buffer timidity-main-buffer)
  (let ((buffer-read-only nil)
	(win-width  (- (window-width) 3))
	(win-height (window-height))
	(str-width  (length (car timidity-demo-string)))
	(str-height (length timidity-demo-string))
	start-point i j k evenp sline c)
    (erase-buffer)
    (if (<= win-height str-height)
	(setq win-height str-height))
    (if (<= win-width str-width)
	(setq win-width str-width))
    (insert-char ?\n win-width)
    (goto-line (/ (- win-height str-height) 2))

    (if (= (% (- win-width str-width) 2) 1)
	(setq win-width (1- win-width)))

    (setq start-point (point)
	  i 0
	  evenp nil)
    (while (< i str-height)
      (if evenp (insert-char ?\  win-width))
      (forward-line 1)
      (setq evenp (not evenp)
	    i (1+ i)))

    (setq j 0)
    (while (< j str-width)
      (goto-char start-point)
      (setq evenp nil
	    i 0
	    sline timidity-demo-string)
      (while (< i str-height)
	(if evenp
	    (progn
	      (delete-char 1)
	      (end-of-line)
	      (insert-char (aref (car sline) j) 1))
	  (insert-char (aref (car sline) (- str-width j 1)) 1))
	(setq i (1+ i)
	      evenp (not evenp)
	      sline (cdr sline))
	(forward-line 1))
      (setq j (1+ j))
      (sit-for 0.01))
    (setq j 0
	  k (/ (- win-width str-width) 2))
    (while (< j k)
      (goto-char start-point)
      (setq evenp nil
	    i 0
	    sline timidity-demo-string)
      (while (< i str-height)
	(if evenp
	    (delete-char 1)
	  (insert-char ?\  1))
	(setq i (1+ i)
	      evenp (not evenp)
	      sline (cdr sline))
	(forward-line 1))
      (setq j (1+ j))
      (sit-for 0.01))))

(mapcar (function (lambda (symbol)
		    (if (byte-code-function-p (symbol-function symbol))
			nil
		      (message "Byte compiling %s..." symbol)
		      (byte-compile symbol))))
	timidity-pre-byte-compiles)

(provide 'timidity)
